parl/agent_zoo/agent_helpers.py

```diff
@@ -0,0 +1,145 @@
+from parl.common.replay_buffer import NoReplacementQueue, ReplayBuffer
+from parl.framework.agent import AgentHelperBase, DataProcessorBase
+
+
+class RLDataProcessor(DataProcessorBase):
+    def __init__(self, specs):
+        super(RLDataProcessor, self).__init__(specs)
+
+    def process_prediction_inputs(self, inputs, states):
+        assert bool(inputs)
+        new_inputs = {}
+        for k in inputs:
+            new_inputs[k] = np.stack([inputs[k]])
+        data = dict(inputs=new_inputs)
+        if bool(states):
+            new_states = {}
+            for k in inputs:
+                new_states[k] = np.stack([new_states[k]])
+            data["states"] = new_states
+        return data
+
+    def process_learning_inputs(self, exp_seqs):
+        def append_data(data, attr, exp_seq):
+            if attr.startswith("next_"):
+                l = [getattr(exp.next_exp, attr[5:]) for exp in exp_seq]
+            else:
+                l = [getattr(exp, attr) for exp in exp_seq]
+            if type(l[0]) is list:
+                data[attr] = data[attr] + l if attr in data else l
+            else:
+                data[attr] = np.concatenate(
+                    [data[attr], np.stack(l)]) if attr in data else np.stack(l)
+
+        data = {}
+        for k in self.specs.iterkeys():
+            data[k] = {}
+            for attr in self.specs[k]:
+                for exp_seq in exp_seqs:
+                    append_data(data[k], attr, exp_seq)
+        return data
+
+
+class OnPolicyHelper(AgentHelperBase):
+    """
+    Example of on-policy helper. It calls learn() every sample_interval steps.
+    While waiting for learning return, the Agent is blocked.
+    """
+
+    def __init__(self, name, communicator, specs, sample_interval=5):
+        super(OnPolicyHelper, self).__init__(name, communicator)
+        self.sample_interval = sample_interval
+        # NoReplacementQueue used to store past experience.
+        self.exp_queue = NoReplacementQueue(sample_seq=False)
+        self.counter = 0
+        self.data_proc = RLDataProcessor(specs)
+
+    def predict(self, inputs, states=dict()):
+        check_type_error(dict, type(inputs))
+        check_type_error(dict, type(states))
+        data = self.data_proc.process_prediction_inputs(inputs, states)
+        self.comm.put_prediction_data(data)
+        ret = self.comm.get_prediction_return()
+        self.counter += 1
+        return ret
+
+    def store_data(self, data):
+        check_type_error(dict, type(data))
+        episode_end = data["episode_end"]
+        self.exp_queue.add(**data)
+        if episode_end or self.counter % self.sample_interval == 0:
+            self.learn()
+
+    def learn(self):
+        self.counter = 0
+        exp_seqs = self.exp_queue.sample()
+        data = self.data_proc.process_learning_inputs(exp_seqs)
+        self.comm.put_training_data(data)
+        self.comm.get_training_return()
+
+
+class ExpReplayHelper(AgentHelperBase):
+    """
```

Comment:
Why this file is put in agent_zoo? It seems to me that they are standard operations and can be fixed.